persistence/sql/persister_session.go

```diff
@@ -76,21 +90,50 @@ func (p *Persister) ListSessionsByIdentity(ctx context.Context, iID uuid.UUID, a
 	return s, nil
 }
 
+// UpsertSession creates a session if not found else updates.
+// This operation also inserts Session device records when a session is being created.
+// The update operation skips updating Session device records since only one record would need to be updated in this case.
 func (p *Persister) UpsertSession(ctx context.Context, s *session.Session) error {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UpsertSession")
 	defer span.End()
 
 	s.NID = p.NetworkID(ctx)
 
-	if err := p.Connection(ctx).Find(new(session.Session), s.ID); errors.Is(err, sql.ErrNoRows) {
-		// This must not be eager or identities will be created / updated
-		return errors.WithStack(p.GetConnection(ctx).Create(s))
-	} else if err != nil {
-		return errors.WithStack(err)
-	}
+	return errors.WithStack(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) error {
+		if err := sqlcon.HandleError(tx.Where("id = ? AND nid = ?", s.ID, s.NID).First(new(session.Session))); errors.Is(err, sqlcon.ErrNoRows) {
+			// This must not be eager or identities will be created / updated
+			if err := sqlcon.HandleError(tx.Create(s)); err != nil {
+				return err
+			}
+
+			for i := range s.Devices {
+				device := &(s.Devices[i])
+				device.SessionID = s.ID
+				device.NID = s.NID
+
+				if device.Location != nil {
+					device.Location = stringsx.GetPointer(stringsx.TruncateByteLen(*device.Location, SessionDeviceLocationMaxLength))
+				}
+				if device.UserAgent != nil {
+					device.UserAgent = stringsx.GetPointer(stringsx.TruncateByteLen(*device.UserAgent, SessionDeviceUserAgentMaxLength))
+				}
 
-	// This must not be eager or identities will be created / updated
-	return p.GetConnection(ctx).Update(s)
+				if err := sqlcon.HandleError(tx.Create(device)); err != nil {
+					return err
+				}
+
+				s.Devices[i] = *device
```

Comment:
I find the pointer handling of `device` a bit confusing here :sweat_smile: 
So you first take the pointer of `s.Devices[i]` and then modify that struct a bit. But that would modify the original struct, so you don't really have to write it back here? Only if you where to do `for i, device := range s.Devices` you would create a copy of the struct and therefore need to write it back here?

Reply:
you're right. I removed the overwrite